pkg/controller/security/security_manager.go

```diff
@@ -0,0 +1,214 @@
+	"kmesh.net/kmesh/pkg/controller/workload"
+	"kmesh.net/kmesh/pkg/logger"
+)
+	
+type Cert struct {
+	hostName string
+	exp time.Time
+}
+
+type PriorityQueue []*Cert
+
+var certsMaps sync.Map
+var saSaveUidNum = make(map[string]int)
+var log = logger.NewLoggerField("security")
+
+type SecretManagerClient struct {
+	caClient *KmeshCaClient
+
+	// configOptions includes all configurable params for the cache.
+	configOptions *istiosecurity.Options
+
+	// storing certificates
+	secretcache *sync.Map
+
+	//caRootPath string
+	pending	PriorityQueue
+}
+ 
+var ScClient SecretManagerClient
+
+const (
+	rootCertPath       = "/var/run/secrets/istio/root-cert.pem"
+)
+   
+func (pq PriorityQueue) Len() int { return len(pq) }
+
+func (pq PriorityQueue) Less(i, j int) bool {
+	return pq[i].exp.Before(pq[j].exp)
+}
+
+func (pq PriorityQueue) Swap(i, j int) {
+	pq[i], pq[j] = pq[j], pq[i]
+}
+
+func (pq *PriorityQueue) Push(x interface{}) {
+	item := x.(*Cert)
+	*pq = append(*pq, item)
+}
+
+func (pq *PriorityQueue) Pop() interface{} {
+	old := *pq
+	n := len(old)
+	item := old[n-1]
+	*pq = old[0 : n-1]
+	return item
+}
+
+// NewSecretManagerClient creates a new SecretManagerClient.
+func NewSecretManagerClient() (error) {
+
+	tlsOpts = &TLSOptions{
+		RootCert:      rootCertPath,
+	}
+
+	options:= NewSecurityOptions()
+	caClient, err := NewKmeshCaClient(options, tlsOpts)
+	if err != nil {
+		return err
+	}
+	
+	pq := make(PriorityQueue, 0)
+	heap.Init(&pq)
+
+	ScClient= SecretManagerClient{
+		caClient:      caClient,
+		configOptions: options,
+		secretcache: &certsMaps,
+		pending:	pq,
+	}
+	go ScClient.refreshExpireCerts()
+	go ScClient.updateCerts()
+	return nil
+}
+ 
+func GetSecretManagerClient() *SecretManagerClient {
+	return &ScClient
+}
+	
+// Automatically check and refresh when the validity period expires
+func (sc *SecretManagerClient) refreshExpireCerts() {
+	var new_certs *istiosecurity.SecretItem
+	var err error
+	var expireTime time.Time
+
+	for {
+		expireTime = time.Now().Add(900 * time.Second)
+		if (len(sc.pending) > 0) {
+			expireTime = sc.pending[0].exp
+		}
+		select{
+		case <-time.After(time.Until(expireTime.Add(-300 * time.Second))):
+			next := heap.Pop(&sc.pending).(*Cert)
+			if _, ok := sc.secretcache.Load(next.hostName); !ok {
+				continue
+			} else {
+				new_certs, err = sc.caClient.fetchCert(next.hostName);
+				if err != nil {
+					log.Errorf("%v refresh fetchCert error : %v", next.hostName, err)
+					return 
+				}
+			}
+			// Check if the key exists in the map, if it does, refresh it, otherwise abandon it. 
+			// If multi-threading refresh errors occur in this round, the certificate will be deleted 
+			// in the next round of workload checks
+			_, ok := sc.secretcache.Load(next.hostName);
+			if ok {
+				sc.secretcache.Store(next.hostName, *new_certs)
+				heap.Push(&sc.pending, &Cert{exp: new_certs.ExpireTime, hostName: new_certs.ResourceName})
+			}
+		case <-time.After(300 * time.Second):
+			continue
```

Comment:
What does this code do?

Reply:
This piece of code checks the most imminent expiration of certificates from the priority queue every 5 minutes, and refreshes them if they are about to expire.